File: src/main/kotlin/ru/lionzxy/tplauncher/utils/UriEncodeUtils.java
package ru.lionzxy.tplauncher.utils;

import java.io.ByteArrayOutputStream;
import java.io.UnsupportedEncodingException;
import java.nio.charset.Charset;

/**
 * Code from Spring Framework
 */
public class UriEncodeUtils {
    public static String encodePath(String source, Charset charset) {
        if (!hasLength(source)) {
            return source;
        }

        byte[] bytes = source.getBytes(charset);
        boolean original = true;
        for (byte b : bytes) {
            if (!isAllowed(b)) {
                original = false;
                break;
            }
        }
        if (original) {
            return source;
        }

        ByteArrayOutputStream baos = new ByteArrayOutputStream(bytes.length);
        for (byte b : bytes) {
            if (isAllowed(b)) {
                baos.write(b);
            } else {
                baos.write('%');
                char hex1 = Character.toUpperCase(Character.forDigit((b >> 4) & 0xF, 16));
                char hex2 = Character.toUpperCase(Character.forDigit(b & 0xF, 16));
                baos.write(hex1);
                baos.write(hex2);
            }
        }
        return copyToString(baos, charset);
    }

    private static boolean isAllowed(int c) {
        return isPchar(c) || '/' == c;
    }

    /**
     * Indicates whether the given character is in the {@code ALPHA} set.
     *
     * @see <a href="https://www.ietf.org/rfc/rfc3986.txt">RFC 3986, appendix A</a>
     */
    private static boolean isAlpha(int c) {
        return (c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z');
    }

    /**
     * Indicates whether the given character is in the {@code DIGIT} set.
     *
     * @see <a href="https://www.ietf.org/rfc/rfc3986.txt">RFC 3986, appendix A</a>
     */
    private static boolean isDigit(int c) {
        return (c >= '0' && c <= '9');
    }

    /**
     * Indicates whether the given character is in the {@code sub-delims} set.
     *
     * @see <a href="https://www.ietf.org/rfc/rfc3986.txt">RFC 3986, appendix A</a>
     */
    private static boolean isSubDelimiter(int c) {
        return ('!' == c || '$' == c || '&' == c || '\'' == c || '(' == c || ')' == c || '*' == c || '+' == c ||
                ',' == c || ';' == c || '=' == c);
    }

    /**
     * Indicates whether the given character is in the {@code unreserved} set.
     *
     * @see <a href="https://www.ietf.org/rfc/rfc3986.txt">RFC 3986, appendix A</a>
     */
    private static boolean isUnreserved(int c) {
        return (isAlpha(c) || isDigit(c) || '-' == c || '.' == c || '_' == c || '~' == c);
    }

    /**
     * Indicates whether the given character is in the {@code pchar} set.
     *
     * @see <a href="https://www.ietf.org/rfc/rfc3986.txt">RFC 3986, appendix A</a>
     */
    private static boolean isPchar(int c) {
        return (isUnreserved(c) || isSubDelimiter(c) || ':' == c || '@' == c);
    }

    /**
     * Check that the given {@code String} is neither {@code null} nor of length 0.
     * <p>Note: this method returns {@code true} for a {@code String} that
     * purely consists of whitespace.
     *
     * @param str the {@code String} to check (may be {@code null})
     * @return {@code true} if the {@code String} is not {@code null} and has length
     */
    private static boolean hasLength(String str) {
        return (str != null && !str.isEmpty());
    }


    /**
     * Copy the contents of the given {@link ByteArrayOutputStream} into a {@link String}.
     * <p>This is a more effective equivalent of {@code new String(baos.toByteArray(), charset)}.
     *
     * @param baos    the {@code ByteArrayOutputStream} to be copied into a String
     * @param charset the {@link Charset} to use to decode the bytes
     * @return the String that has been copied to (possibly empty)
     * @since 5.2.6
     */
    private static String copyToString(ByteArrayOutputStream baos, Charset charset) {
        try {
            // Can be replaced with toString(Charset) call in Java 10+
            return baos.toString(charset.name());
        } catch (UnsupportedEncodingException ex) {
            // Should never happen
            throw new IllegalArgumentException("Invalid charset name: " + charset, ex);
        }
    }
}

File: core/src/main/kotlin/ru/lionzxy/tplauncher/utils/UriEncodeUtils.kt
package ru.lionzxy.tplauncher.utils

import java.io.ByteArrayOutputStream
import java.nio.charset.Charset

/**
 * Code from Spring Framework.
 */
object UriEncodeUtils {

    fun encodePath(source: String, charset: Charset): String {
        if (!hasLength(source)) {
            return source
        }

        val bytes = source.toByteArray(charset)
        var original = true
        for (b in bytes) {
            if (!isAllowed(b.toInt())) {
                original = false
                break
            }
        }
        if (original) {
            return source
        }

        val baos = ByteArrayOutputStream(bytes.size)
        for (b in bytes) {
            if (isAllowed(b.toInt())) {
                baos.write(b.toInt())
            } else {
                baos.write('%'.code)
                val hex1 = Character.toUpperCase(Character.forDigit((b.toInt() shr 4) and 0xF, 16))
                val hex2 = Character.toUpperCase(Character.forDigit(b.toInt() and 0xF, 16))
                baos.write(hex1.code)
                baos.write(hex2.code)
            }
        }
        return baos.toString(charset)
    }

    private fun isAllowed(c: Int): Boolean = isPchar(c) || '/'.code == c

    /** Whether [c] is in the `ALPHA` set (RFC 3986, appendix A). */
    private fun isAlpha(c: Int): Boolean =
        (c >= 'a'.code && c <= 'z'.code) || (c >= 'A'.code && c <= 'Z'.code)

    /** Whether [c] is in the `DIGIT` set (RFC 3986, appendix A). */
    private fun isDigit(c: Int): Boolean = c >= '0'.code && c <= '9'.code

    /** Whether [c] is in the `sub-delims` set (RFC 3986, appendix A). */
    private fun isSubDelimiter(c: Int): Boolean =
        '!'.code == c || '$'.code == c || '&'.code == c || '\''.code == c || '('.code == c ||
            ')'.code == c || '*'.code == c || '+'.code == c || ','.code == c || ';'.code == c ||
            '='.code == c

    /** Whether [c] is in the `unreserved` set (RFC 3986, appendix A). */
    private fun isUnreserved(c: Int): Boolean =
        isAlpha(c) || isDigit(c) || '-'.code == c || '.'.code == c || '_'.code == c || '~'.code == c

    /** Whether [c] is in the `pchar` set (RFC 3986, appendix A). */
    private fun isPchar(c: Int): Boolean = isUnreserved(c) || isSubDelimiter(c) || ':'.code == c || '@'.code == c

    /** True when [str] is neither null nor empty (whitespace counts as non-empty). */
    private fun hasLength(str: String?): Boolean = str != null && str.isNotEmpty()
}
